File: stack-queues/2390.py
# Tc = O(n)
# Sc = O(n)


# Approach-1
class Solution:
    def removeStars(self, s: str) -> str:
        st = []
        for char in s:
            if char == '*':
                if st:
                    st.pop()
            else:
                st.append(char)
        result = ''.join(st)
        return result
    
# Approach-2
class Solution:
    def removeStars(self, s: str) -> str:
        res = []
        for char in s:
            if char == '*':
                    res.pop()
            else:
                res.append(char)
        return ''.join(res)

# Approach-3
class Solution:
    def removeStars(self, s: str) -> str:
        n = len(s)
        temp = [''] * n
        j = 0
        for i in range(n):
            if s[i] == '*':
                j -= 1  
            else:
                temp[j] = s[i]  
                j += 1  
        result = ''.join(temp[:j])
        return result

File: ADT/stack-queue-pblm/2390.cpp
// Microsoft

// stack
class Solution {
public:
    string removeStars(string s) {
        stack<char> st;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '*')
                st.pop();
            else
                st.push(s[i]);
        }
        string result = "";
        while (!st.empty()) {
            result.push_back(st.top());
            st.pop();
        }
        reverse(begin(result), end(result));
        return result;
    }
};

// string
class Solution {
public:
    string removeStars(string s) {
        string result = "";
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '*')
                result.pop_back();
            else
                result.push_back(s[i]);
        }
        return result;
    }
};


// 2 pointers
class Solution {
public:
    string removeStars(string s) {
        vector<char> ch(s.size());
        int j = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '*') {
                j--;
            } else {
                ch[j++] = s[i];
            }
        }

        string result = "";
        for (int i = 0; i < j; i++) {
            result.push_back(ch[i]);
        }

        return result;
    }
};
